File: frontend/src/api/ai.ts
import apiClient from './client';
import type {
  AIDigest,
  AIRisksResponse,
  AIOverloadEntry,
  AIParsedTask,
  AIAssigneeCandidate,
  AIGoalSummary,
} from '@/types';

export async function getDigest(params: {
  scope: string;
  id?: string;
  period: string;
}): Promise<AIDigest> {
  const res = await apiClient.post<AIDigest>('/ai/digest', params);
  return res.data;
}

export async function getRisks(params: {
  scope: string;
  id?: string;
}): Promise<AIRisksResponse> {
  const res = await apiClient.post('/ai/risks', params);
  const raw = (res.data?.items ?? []) as any[];
  const items = raw.map((it) => ({
    task_id: it.task_id,
    task_title: it.task_title ?? it.title ?? '',
    risk_level: (it.risk_level === 'med' ? 'medium' : it.risk_level) as 'high' | 'medium',
    reason: it.reason ?? '',
  }));
  return { items };
}

export async function getOverload(): Promise<{ items: AIOverloadEntry[] }> {
  const res = await apiClient.post<{ items: AIOverloadEntry[] }>('/ai/overload');
  return res.data;
}

export async function parseTask(text: string): Promise<AIParsedTask> {
  const res = await apiClient.post<AIParsedTask>('/ai/parse-task', { text });
  return res.data;
}

export async function suggestAssignee(params: {
  title: string;
  description?: string;
  department_id?: string;
  priority?: string;
  due_date?: string;
}): Promise<{ candidates: AIAssigneeCandidate[] }> {
  const res = await apiClient.post<{ candidates: AIAssigneeCandidate[] }>(
    '/ai/suggest-assignee',
    params
  );
  return res.data;
}

export async function chat(params: {
  message: string;
  context?: { task_id?: string; department_id?: string };
  history?: Array<{ role: 'user' | 'assistant'; content: string }>;
}): Promise<{ reply: string }> {
  const res = await apiClient.post<{ reply: string }>('/ai/chat', params);
  return res.data;
}

export interface AssistantButton {
  label: string;
  callback_data: string;
}

export interface AssistantMessageResponse {
  conversation_id: string;
  message_id?: string;
  text: string;
  buttons: AssistantButton[];
  pending_action?: Record<string, unknown> | null;
  referenced_task_ids: string[];
  mode: 'normal' | 'degraded';
  model: {
    provider: string;
    name: string;
    available: boolean;
  };
}

export interface AssistantConversationResponse {
  conversation_id: string;
  messages: Array<{ id: string; role: 'user' | 'assistant'; content: string }>;
  mode: 'normal' | 'degraded';
  model: {
    provider: string;
    name: string;
    available: boolean;
  };
}

export async function getAssistantConversation(conversationId?: string): Promise<AssistantConversationResponse> {
  const res = await apiClient.get<AssistantConversationResponse>('/ai/assistant/conversation', {
    params: conversationId ? { conversation_id: conversationId } : undefined,
  });
  return res.data;
}

export async function sendAssistantMessage(params: {
  message: string;
  conversation_id?: string;
  context?: { task_id?: string; department_id?: string };
}): Promise<AssistantMessageResponse> {
  const res = await apiClient.post<AssistantMessageResponse>('/ai/assistant/message', params);
  return res.data;
}

export async function confirmAssistantAction(actionId: string): Promise<{ text: string; buttons: AssistantButton[] }> {
  const res = await apiClient.post<{ text: string; buttons: AssistantButton[] }>(
    `/ai/assistant/actions/${actionId}/confirm`
  );
  return res.data;
}

export async function cancelAssistantAction(actionId: string): Promise<{ text: string; buttons: AssistantButton[] }> {
  const res = await apiClient.post<{ text: string; buttons: AssistantButton[] }>(
    `/ai/assistant/actions/${actionId}/cancel`
  );
  return res.data;
}

export async function getGoalSummary(goalId: string): Promise<AIGoalSummary> {
  const res = await apiClient.post<AIGoalSummary>('/ai/goal-summary', { goal_id: goalId });
  return res.data;
}

File: backend/app/api/ai.py
import uuid
from typing import Optional

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.deps import get_db, get_current_user
from app.models.user import User
from app.schemas.ai import (
    DigestRequest, DigestResponse,
    RisksRequest, RisksResponse,
    OverloadResponse,
    ParseTaskRequest, ParseTaskResponse,
    SuggestAssigneeRequest, SuggestAssigneeResponse,
    ChatRequest, ChatResponse,
    GoalSummaryResponse,
    AssistantMessageRequest, AssistantMessageResponse,
    AssistantConversationResponse, AssistantConfirmResponse,
    AssistantStoredMessage,
)
from app.services import ai_service
from app.services import assistant_agent

router = APIRouter(prefix="/ai", tags=["ai"])


@router.post("/digest", response_model=DigestResponse)
async def ai_digest(body: DigestRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await ai_service.digest(db, body.scope, body.id, body.period or "month", current_user)
    return result


@router.post("/risks", response_model=RisksResponse)
async def ai_risks(body: RisksRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await ai_service.risks(db, body.scope, body.id, current_user)
    return result


@router.post("/overload", response_model=OverloadResponse)
async def ai_overload(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await ai_service.overload(db, current_user)
    return result


@router.post("/parse-task", response_model=ParseTaskResponse)
async def ai_parse_task(body: ParseTaskRequest, current_user: User = Depends(get_current_user)):
    result = await ai_service.parse_task(body.text)
    return result


@router.post("/suggest-assignee", response_model=SuggestAssigneeResponse)
async def ai_suggest_assignee(
    body: SuggestAssigneeRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await ai_service.suggest_assignee(
        db, body.title, body.description, body.department_id, body.priority, body.due_date, current_user,
    )
    return result


@router.post("/chat", response_model=ChatResponse)
async def ai_chat(body: ChatRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await assistant_agent.process_message(
        db,
        current_user,
        body.message,
        channel="web",
        context=body.context or {},
    )
    return {"reply": result.text}


@router.get("/assistant/conversation", response_model=AssistantConversationResponse)
async def assistant_conversation(
    conversation_id: Optional[uuid.UUID] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    conversation, messages = assistant_agent.get_conversation_messages(db, current_user, conversation_id, channel="web")
    return {
        "conversation_id": conversation.id,
        "messages": [AssistantStoredMessage(id=item.id, role=item.role, content=item.content) for item in messages],
        "mode": "degraded" if conversation.llm_status == "down" else "normal",
        "model": {
            "provider": assistant_agent.settings.LLM_PROVIDER,
            "name": assistant_agent._model_name(),
            "available": conversation.llm_status == "up",
        },
    }


@router.post("/assistant/message", response_model=AssistantMessageResponse)
async def assistant_message(
    body: AssistantMessageRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await assistant_agent.process_message(
        db,
        current_user,
        body.message,
        channel="web",
        conversation_id=body.conversation_id,
        external_chat_id="default",
        context=body.context or {},
    )
    return {
        "conversation_id": result.conversation_id,
        "message_id": result.message_id,
        "text": result.text,
        "buttons": result.buttons,
        "pending_action": result.pending_action,
        "referenced_task_ids": result.referenced_task_ids,
        "mode": result.mode,
        "model": result.model,
    }


@router.delete("/assistant/conversation", response_model=AssistantConversationResponse)
async def assistant_clear_conversation(
    conversation_id: Optional[uuid.UUID] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    conversation = assistant_agent.clear_conversation(db, current_user, conversation_id, channel="web", external_chat_id="default")
    return {
        "conversation_id": conversation.id,
        "messages": [],
        "mode": "degraded" if conversation.llm_status == "down" else "normal",
        "model": {
            "provider": assistant_agent.settings.LLM_PROVIDER,
            "name": assistant_agent._model_name(),
            "available": conversation.llm_status == "up",
        },
    }


@router.post("/assistant/actions/{action_id}/confirm", response_model=AssistantConfirmResponse)
async def assistant_confirm_action(
    action_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await assistant_agent.confirm_action(db, current_user, action_id, confirm=True)
    return {"text": result.text, "buttons": result.buttons, "referenced_task_ids": result.referenced_task_ids}


@router.post("/assistant/actions/{action_id}/cancel", response_model=AssistantConfirmResponse)
async def assistant_cancel_action(
    action_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await assistant_agent.confirm_action(db, current_user, action_id, confirm=False)
    return {"text": result.text, "buttons": result.buttons, "referenced_task_ids": result.referenced_task_ids}


@router.post("/goal-summary", response_model=GoalSummaryResponse)
async def ai_goal_summary(
    body: dict,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    goal_id = uuid.UUID(str(body.get("goal_id")))
    result = await ai_service.goal_summary(db, goal_id, current_user)
    return result
